py/obj.h

```diff
@@ -664,9 +664,29 @@ extern const struct _mp_obj_exception_t mp_const_GeneratorExit_obj;
 // check for more specific object types.
 // Note: these are kept as macros because inline functions sometimes use much
 // more code space than the equivalent macros, depending on the compiler.
-#define mp_obj_is_type(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t*)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
-#define mp_obj_is_int(o) (mp_obj_is_small_int(o) || mp_obj_is_type(o, &mp_type_int))
-#define mp_obj_is_str(o) (mp_obj_is_qstr(o) || mp_obj_is_type(o, &mp_type_str))
+
+// don't use mp_obj_is_type_unsafe directly; use mp_obj_is_type which provides additional safety checks.
+// use the former only if you need to bypass these checks (because you've already checked everything)
+#define mp_obj_is_type_unsafe(o, t) (mp_obj_is_obj(o) && (((mp_obj_base_t*)MP_OBJ_TO_PTR(o))->type == (t))) // this does not work for checking int, str or fun; use below macros for that
+// type checks are split to a separate, constant result macro. I found this way to work without hindering GCC's
+// optimizations (other tricks like using ({ expr; exper; }) or (exp, expr, expr) in mp_obj_is_type result in
+// missed optimizations. however, this way GCC retains the optimizations it used to have before these checks
+// were added).
+#define _type_checks(t) (                                                          \
+     MP_STATIC_ASSERT((t) != &mp_type_bool), assert((t) != &mp_type_bool),         \
```

Comment:
Doesn't work for msvc: static assert requires a constant expression.. You could leave them out `#ifdef _MSC_VER`. Since this is part of the public API I'd also opt for a more dedicated name prefixed with `mp_`, 'type_checks' is pretty generic/common, something like `mp_obj_is_type_checks`

Reply:
But clobbering the global namespace with names from MPY is not okay either. I guess `mp_obj_is_type_checks` is good enough. Perhaps even `mp_obj_is_type_static_checks`.